images/virtualization-artifact/pkg/controller/service/ip_address_service.go

```diff
@@ -28,98 +29,95 @@ import (
 )
 
 type IpAddressService struct {
-	logger      logr.Logger
-	ParsedCIDRs []*net.IPNet
+	parsedCIDRs []netip.Prefix
 }
 
 func NewIpAddressService(
 	logger logr.Logger,
 	virtualMachineCIDRs []string,
 ) *IpAddressService {
-	parsedCIDRs := make([]*net.IPNet, len(virtualMachineCIDRs))
+	parsedCIDRs := make([]netip.Prefix, len(virtualMachineCIDRs))
 
 	for i, cidr := range virtualMachineCIDRs {
-		_, parsedCIDR, err := net.ParseCIDR(cidr)
-		if err != nil || parsedCIDR == nil {
+		parsedCIDR, err := netip.ParsePrefix(cidr)
+		if err != nil {
 			logger.Error(err, fmt.Sprintf("failed to parse CIDR %s:", cidr), "err", err)
 			return nil
 		}
 		parsedCIDRs[i] = parsedCIDR
 	}
 
 	return &IpAddressService{
-		logger:      logger.WithValues("service", "IpAddressService"),
-		ParsedCIDRs: parsedCIDRs,
+		parsedCIDRs: parsedCIDRs,
 	}
 }
 
-func (s IpAddressService) IsAvailableAddress(address string, allocatedIPs common.AllocatedIPs) bool {
-	ip := net.ParseIP(address)
+func (s IpAddressService) IsAvailableAddress(address string, allocatedIPs common.AllocatedIPs) error {
+	ip, err := netip.ParseAddr(address)
+	if err != nil || !ip.IsValid() {
+		return errors.New("invalid IP address format")
+	}
 
 	if _, ok := allocatedIPs[ip.String()]; !ok {
-		for _, cidr := range s.ParsedCIDRs {
+		for _, cidr := range s.parsedCIDRs {
 			if cidr.Contains(ip) {
 				// available
-				return true
+				return nil
 			}
 		}
 		// out of range
-		return false
+		return ErrIPAddressOutOfRange
 	}
+
 	// already exists
-	return false
+	return ErrIPAddressAlreadyExist
 }
 
 func (s IpAddressService) AllocateNewIP(allocatedIPs common.AllocatedIPs) (string, error) {
-	for _, cidr := range s.ParsedCIDRs {
-		for ip := cidr.IP.Mask(cidr.Mask); cidr.Contains(ip); inc(ip) {
-			// Allow allocation of IP address explicitly specified using a 32-bit mask.
-			if k8snet.RangeSize(cidr) != 1 {
-				// Skip the allocation of the first or last addresses within the CIDR range.
+	for _, cidr := range s.parsedCIDRs {
+		for ip := cidr.Addr(); cidr.Contains(ip); ip = ip.Next() {
+			if k8snet.RangeSize(toIPNet(cidr)) != 1 {
 				isFirstLast, err := isFirstLastIP(ip, cidr)
 				if err != nil {
 					return "", err
 				}
-
 				if isFirstLast {
 					continue
 				}
 			}
 
-			_, ok := allocatedIPs[ip.String()]
-			if !ok {
+			if _, ok := allocatedIPs[ip.String()]; !ok {
 				return ip.String(), nil
 			}
 		}
 	}
 	return "", errors.New("no remaining ips")
 }
 
-func inc(ip net.IP) {
-	for j := len(ip) - 1; j >= 0; j-- {
-		ip[j]++
-		if ip[j] > 0 {
-			break
-		}
+func toIPNet(prefix netip.Prefix) *net.IPNet {
+	return &net.IPNet{
+		IP:   prefix.Masked().Addr().AsSlice(),
+		Mask: net.CIDRMask(prefix.Bits(), prefix.Addr().BitLen()),
 	}
 }
 
-func isFirstLastIP(ip net.IP, cidr *net.IPNet) (bool, error) {
-	size := int(k8snet.RangeSize(cidr))
+func isFirstLastIP(ip netip.Addr, cidr netip.Prefix) (bool, error) {
```

Comment:
It'll be great to add simple unit tests for isFirstLastIP and toIPNet and for service public methods.

Reply:
I added tests.